packages/tpetra/core/src/Tpetra_Details_residual.hpp

```diff
@@ -0,0 +1,280 @@
+  // We don't try to to detect the more subtle cases (e.g., one is a
+  // subview of the other, but their initial pointers differ).  We
+  // only need to do this if this matrix's Import is trivial;
+  // otherwise, we don't actually apply the operator from X into Y.
+  
+  RCP<const import_type> importer = A.getGraph ()->getImporter ();
+  RCP<const export_type> exporter = A.getGraph ()->getExporter ();
+
+  // Temporary MV for Import operation.  After the block of code
+  // below, this will be an (Imported if necessary) column Map MV
+  // ready to give to localApply(...).
+  RCP<const MV> X_colMap;
+  if (importer.is_null ()) {
+    if (! X_in.isConstantStride ()) {
+      // Not all sparse mat-vec kernels can handle an input MV with
+      // nonconstant stride correctly, so we have to copy it in that
+      // case into a constant stride MV.  To make a constant stride
+      // copy of X_in, we force creation of the column (== domain)
+      // Map MV (if it hasn't already been created, else fetch the
+      // cached copy).  This avoids creating a new MV each time.
+      RCP<MV> X_colMapNonConst = A.getColumnMapMultiVector (X_in, true);
+      Tpetra::deep_copy (*X_colMapNonConst, X_in);
+      X_colMap = rcp_const_cast<const MV> (X_colMapNonConst);
+    }
+    else {
+      // The domain and column Maps are the same, so do the local
+      // multiply using the domain Map input MV X_in.
+      X_colMap = rcpFromRef (X_in);
+    }
+  }
+  else { // need to Import source (multi)vector
+    ProfilingRegion regionImport ("Tpetra::CrsMatrix::residual: Import");
+    // We're doing an Import anyway, which will copy the relevant
+    // elements of the domain Map MV X_in into a separate column Map
+    // MV.  Thus, we don't have to worry whether X_in is constant
+    // stride.
+    RCP<MV> X_colMapNonConst = A.getColumnMapMultiVector (X_in);
+    
+    // Import from the domain Map MV to the column Map MV.
+    X_colMapNonConst->doImport (X_in, *importer, INSERT);
+    X_colMap = rcp_const_cast<const MV> (X_colMapNonConst);
+  }
+
+  // Get a vector for the rowMap output residual
+  RCP<MV> R_rowMap;
+  if(exporter.is_null()) {
+    R_rowMap = rcpFromRef(R_in);
+  }
+  else {
+    R_rowMap = A.getRowMapMultiVector (R_in);    
+  }
+  
+  // FIXME: Need to get a potential cached buffer for B (this will need to be another function)
+  RCP<const MV> B_rowMap =  rcpFromRef(B_in); // A.getRowMapMultiVector (B_in);
+
+
+  // If we have a nontrivial Export object, we must perform an
+  // Export.  In that case, the local multiply result will go into
+  // the row Map multivector.  We don't have to make a
+  // constant-stride version of R_in in this case, because we had to
+  // make a constant stride R_rowMap MV and do an Export anyway.
+  if (! exporter.is_null ()) {
+    {
+      ProfilingRegion regionExport ("Tpetra::CrsMatrix::residual: B Import");
+      //      B_rowMap->doImport(B_in, *exporter, ADD);//FIXME
```

Comment:
Do this, and also handle nonconstant stride B.

Reply:
Got it